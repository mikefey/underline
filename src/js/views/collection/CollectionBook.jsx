import React from 'react';
import { Link } from 'react-router';
import stringHelper from 'libs/string/string-helper';
import style from './CollectionBook.scss';


/**
 * Component for each book in the collection
 */
const CollectionBook = (props) => {
  /**
   * Returns max length for title and author text
   * @returns {Number} - Max length for title and author text
   */
  const getTextMaxLength = () => {
    const windowWidth = props.windowSize.width;
    let maxLength = 18;

    if (windowWidth > 320) {
      maxLength = 22;
    }

    if (windowWidth >= 400) {
      maxLength = 26;
    }

    if (windowWidth >= 500) {
      maxLength = 34;
    }

    if (windowWidth >= 600) {
      maxLength = 44;
    }

    if (windowWidth >= 700) {
      maxLength = 54;
    }

    if (windowWidth >= 768) {
      maxLength = 60;
    }

    return maxLength;
  };


  /**
   * Renders cover image or placeholder if no cover exists
   * @returns {ReactElement} - A cover image or placeholder
   */
  const renderCover = () => {
    let element;

    if (props.data.cover_image && props.data.cover_image.length) {
      element = (
        <img
          alt={props.data.title}
          src={props.data.cover_image}
        />
      );
    } else {
      const firstLetter = props.data.title && props.data.title.length
        ? props.data.title[0].toUpperCase()
        : '';
      element = (
        <div className={style.imagePlaceholder}>{firstLetter}</div>
      );
    }

    return element;
  };


  /**
   * Click handler for 'remove' button
   * @returns {undefined} undefined
   */
  const onRemoveButtonClick = () => {
    props.removeButtonClickHandler(props.data.id);
  };

  const textMaxLength = getTextMaxLength();
  let title = props.data.title;
  let author = props.data.author;

  if (title.length > textMaxLength) {
    title = stringHelper.truncate(title, textMaxLength - 3, '...');
  }

  if (author.length > textMaxLength) {
    author = stringHelper.truncate(author, textMaxLength - 3, '...');
  }

  const progressStyle = {
    width: `${props.data.current_location.percentage * 100}%`,
  };

  return (
    <div className={style.wrapper}>
      <Link to={`/book/${props.data.id}`} className={style.button}>
        <div className={style.imageWrapper}>
          {renderCover()}
        </div>
        <div className={style.textWrapper}>
          <h2 className={style.title}>{title}</h2>
          <div className={style.author}>{author}</div>
        </div>
        <div className={style.progressWrapper}>
          <div className={style.progressBack} />
          <div style={progressStyle} className={style.progress} />
        </div>
      </Link>
      <button
        className={style.removeButton}
        onClick={onRemoveButtonClick}
      >
        <div className={style.removeButtonIcon}>
          <div className={style.removeButtonIconLine} />
          <div className={style.removeButtonIconLineTwo} />
        </div>
      </button>
    </div>
  );
};


/**
 * PropTypes
 * @prop {Object} data - An object with a book's metadata
 * @prop {Function} removeButtonClickHandler - A click handler for the
 * @prop {Object} windowSize - An object holding the width and height of the
 * window
 */
CollectionBook.propTypes = {
  data: React.PropTypes.object.isRequired,
  removeButtonClickHandler: React.PropTypes.func.isRequired,
  windowSize: React.PropTypes.object.isRequired,
};

export default CollectionBook;
