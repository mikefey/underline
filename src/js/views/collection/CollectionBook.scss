@import '~colors.scss';

.wrapper {
  border-bottom: 1px solid $dark-grey;
  float: left;
  height: 100px;
  position: relative;
  text-align: left;
  width: 100%;
}

.button {
  color: $dark-grey;
  display: block;
  height: 100%;
  width: 100%;
}

.removeButton {
  bottom: 0;
  cursor: pointer;
  height: 40px;
  position: absolute;
  right: 10px;
  width: 40px;
}

.removeButtonIcon {
  background-color: $red;
  border-radius: 50%;
  font-size: 10px;
  font-weight: 600;
  height: 15px;
  margin: 11px 0 0 12px;
  position: absolute;
  width: 15px;
  top: 0;
}

.removeButtonIconLine {
  background-color: $white;
  height: 1px;
  margin: 7px 0 0 3px;
  position: absolute;
  transform: rotate(-45deg);
  width: 9px;
}

.removeButtonIconLineTwo {
  composes: removeButtonIconLine;
  transform: rotate(45deg);
}

.imageWrapper {
  float: left;
  height: 100%;
  overflow: hidden;
  padding: 10px;
  position: relative;
  width: 30%;

  @media(min-width: 600px) {
    width: 20%;
  }

  img {
    box-shadow: 0 0 10px 1px $medium-opacity-black;
    float: left;
    left: 50%;
    max-height: calc(100% - 20px);
    max-width: calc(100% - 20px);
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%);
  }
}

.imagePlaceholder {
  background-color: $light-grey;
  border: 1px solid $dark-grey;
  box-shadow: 0 0 10px 1px rgba(0, 0, 0, 0.2);
  float: left;
  height: 100%;
  left: 50%;
  max-height: calc(100% - 20px);
  max-width: 50%;
  padding-top: 25px;
  position: absolute;
  text-align: center;
  top: 50%;
  transform: translate(-50%, -50%);
  width: 100%;
}

.textWrapper {
  float: left;
  padding: 10px;
  position: relative;
  width: 70%;

  @media(min-width: 600px) {
    width: 80%;
  }
}

.title {
  float: left;
  line-height: 18px;
  position: relative;
  font-size: 18px;
  width: 100%;
}

.author {
  float: left;
  font-size: 16px;
  line-height: 16px;
  margin-top: 10px;
  position: relative;
  width: 100%;
}

.progressWrapper {
  bottom: 15px;
  height: 2px;
  left: calc(30% + 10px);
  position: absolute;
  width: calc(70% - 70px);

  @media(min-width: 600px) {
    left: calc(20% + 10px);
    width: calc(80% - 70px);
  }
}

.progressBack {
  background-color: $light-grey;
  height: 100%;
  position: absolute;
  width: 100%;
}

.progress {
  background-color: $rose;
  height: 100%;
  opacity: 0.7;
  position: absolute;
  width: 100%;
}
