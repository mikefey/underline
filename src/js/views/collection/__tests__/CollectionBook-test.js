import test from 'ava';
import React from 'react';
import { renderToString } from 'react-dom/server';
import CollectionBook from './../CollectionBook.jsx';
import bookFixture from './../../../../../test/fixtures/book';

test('CollectionBook component', (t) => {
  const windowSize = {
    width: 320,
    height: 500,
  };

  const removeButtonClickHandler = () => {};
  const updateCurrentBookData = () => {};

  const tree = renderToString(<CollectionBook
    data={bookFixture}
    windowSize={windowSize}
    removeButtonClickHandler={removeButtonClickHandler}
    updateCurrentBookData={updateCurrentBookData}
  />);

  t.snapshot(tree);
});
