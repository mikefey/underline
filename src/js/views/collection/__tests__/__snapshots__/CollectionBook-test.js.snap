// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionBook component 1`] = `"<div class=\\"CollectionBook__wrapper___3GBcv3hJrHNyAAL\\" data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"722914099\\"><a class=\\"CollectionBook__button___1AbKlIBSykTTsGD\\" data-reactid=\\"2\\"><div class=\\"CollectionBook__imageWrapper___y3SmxAA224b7t7M\\" data-reactid=\\"3\\"><img alt=\\"A Good Book\\" src=\\"data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7\\" data-reactid=\\"4\\"/></div><div class=\\"CollectionBook__textWrapper___2phBcF6leMT_Szd\\" data-reactid=\\"5\\"><h2 class=\\"CollectionBook__title___32U5TK0ztfAD8ec\\" data-reactid=\\"6\\">A Good Book</h2><div class=\\"CollectionBook__author___3Es49tS7ViSMldX\\" data-reactid=\\"7\\">The book author</div></div><div class=\\"CollectionBook__progressWrapper___3edHqIX5PYR9iz3\\" data-reactid=\\"8\\"><div class=\\"CollectionBook__progressBack___w9pyWwm5HWftZxm\\" data-reactid=\\"9\\"></div><div style=\\"width:0%;\\" class=\\"CollectionBook__progress___2sOJs1jf_xQxl5S\\" data-reactid=\\"10\\"></div></div></a><button class=\\"CollectionBook__removeButton___3Yv0v9TRZegontQ\\" data-reactid=\\"11\\"><div class=\\"CollectionBook__removeButtonIcon___2P-RDWN-az8BBbv\\" data-reactid=\\"12\\"><div class=\\"CollectionBook__removeButtonIconLine___Z6gOfidQeF49xXl\\" data-reactid=\\"13\\"></div><div class=\\"CollectionBook__removeButtonIconLineTwo___1FPu7yMSK8_gynN CollectionBook__removeButtonIconLine___Z6gOfidQeF49xXl\\" data-reactid=\\"14\\"></div></div></button></div>"`;
